router/status.go

```diff
@@ -0,0 +1,206 @@
+	Name        string
+	NickName    string
+	UID         PeerUID
+	Version     uint64
+	Connections []ConnectionStatus
+}
+
+type ConnectionStatus struct {
+	Name        string
+	NickName    string
+	Address     string
+	Outbound    bool
+	Established bool
+}
+
+type UnicastRouteStatus struct {
+	Dest, Via string
+}
+
+type BroadcastRouteStatus struct {
+	Source string
+	Via    []string
+}
+
+func NewStatus(router *Router) *Status {
+	return &Status{
+		router.UsingPassword(),
+		router.PeerDiscovery,
+		router.Ourself.Name.String(),
+		router.Ourself.NickName,
+		router.Port,
+		router.Iface,
+		NewMACStatusSlice(router.Macs),
+		NewPeerStatusSlice(router.Peers),
+		NewUnicastRouteStatusSlice(router.Routes),
+		NewBroadcastRouteStatusSlice(router.Routes),
+		NewConnectionMakerStatus(router.ConnectionMaker)}
+}
+
+func NewConnectionMakerStatus(cm *ConnectionMaker) ConnectionMakerStatus {
+	// We need to Refresh first in order to clear out any 'attempting'
+	// connections from cm.targets that have been established since
+	// the last run of cm.checkStateAndAttemptConnections. These
+	// entries are harmless but do represent stale state that we do
+	// not want to report.
+	cm.Refresh()
+	resultChan := make(chan ConnectionMakerStatus, 0)
+	cm.actionChan <- func() bool {
+		status := ConnectionMakerStatus{
+			DirectPeers: []string{},
+			Reconnects:  []TargetStatus{},
+		}
+		for peer := range cm.directPeers {
+			status.DirectPeers = append(status.DirectPeers, peer)
+		}
+
+		for address, target := range cm.targets {
+			status.Reconnects = append(status.Reconnects, newTargetStatus(address, target))
+		}
+		resultChan <- status
+		return false
+	}
+	return <-resultChan
+}
+
+func newTargetStatus(address string, target *Target) TargetStatus {
+	status := TargetStatus{
+		Address:     address,
+		Attempting:  target.attempting,
+		TryAfter:    target.tryAfter,
+		TryInterval: target.tryInterval.String(),
+	}
+	if target.lastError != nil {
+		status.LastError = target.lastError.Error()
+	}
+	return status
+}
+
+func NewMACStatusSlice(cache *MacCache) []MACStatus {
+	cache.RLock()
+	defer cache.RUnlock()
+
+	var slice []MACStatus
+	for key, entry := range cache.table {
+		slice = append(slice,
+			MACStatus{
+				intmac(key).String(),
+				entry.peer.Name.String(),
+				entry.peer.NickName,
+				entry.lastSeen})
+	}
+
+	return slice
+}
+
+func NewPeerStatusSlice(peers *Peers) []PeerStatus {
+	var slice []PeerStatus
+
+	peers.ForEach(func(peer *Peer) {
+		var connections []ConnectionStatus
+		if peer == peers.ourself.Peer {
+			for conn := range peers.ourself.Connections() {
+				connections = append(connections, newConnectionStatus(conn))
+			}
+		} else {
+			// Modifying peer.connections requires a write lock on
+			// Peers, and since we are holding a read lock (due to the
+			// ForEach), access without locking the peer is safe.
+			for _, conn := range peer.connections {
+				connections = append(connections, newConnectionStatus(conn))
+			}
+		}
```

Comment:
Hmm that's not straightforward - `peers.ourself.Connections()` is of type `map[Connection]struct{}`, whereas `peer.connections` is of type `map[PeerName]Connection`...